File: IC2/sample/check/check.m
% for i=1:100 % get 100 pictures
i=5;
%% open document
fdata=fopen('./sample/data.txt','r');
fweight=fopen('./sample/weight.txt','r');
foutput=fopen('./sample/output.txt','r');

%% get original data
C_data=textscan(fdata,'%bs8');
M_data=C_data{:};
data_matrix=reshape(M_data(((i-1)*64+1):(i*64)),[8,8]);
data_matrix=data_matrix';

C_weight=textscan(fweight,'%bs8');
M_weight=C_weight{1};
weight_conv=reshape(M_weight(1:3*3*3),[3,3,3]);
weight_conv=permute(weight_conv,[2,1,3]);
weight_connect=reshape(M_weight(3*3*3+1:end),[3,3,3]);
weight_connect=permute(weight_connect,[2,1,3]);

C_output=textscan(foutput,'%bs8');
output_check=C_output{:};
output_check=output_check(i);

%% cauculate convolution
conv_ans=int32(zeros(6,6,3));
for d=1:3
    for r=1:6
        for c=1:6
            conv_ans(r,c,d)=int32(sum(int32(data_matrix(r:(r+2),c:(c+2))).*int32(weight_conv(:,:,d)),'all'));
        end
    end
end

%% compress bits
conv_compress=bitshift(int32(conv_ans),-8,"int32");
conv_compress=int8(conv_compress>127)*int8(127)+...
    int8(conv_compress<-128)*int8(-128)+...
    int8(conv_compress>=-128 & conv_compress<=127).*int8(conv_compress);

%% calculate pooling
pool_ans=int8(zeros(3,3,3));
for d=1:3
    for r=1:3
        for c=1:3
            pool_ans(r,c,d)=max(conv_compress((r*2-1):(r*2),(c*2-1):(c*2),d),[],'all');
        end
    end
end

%% calculate connect
connect_ans=int32(sum(int32(pool_ans(:,:,:)).*int32(weight_connect(:,:,:)),'all'));

%% compress bits
connect_compress=bitshift(int32(connect_ans),-8,"int32");
connect_compress=int8(connect_compress>127)*int8(127)+...
    int8(connect_compress<-128)*int8(-128)+...
    int8(connect_compress>=-128 & connect_compress<=127).*int8(connect_compress);

%% output check
if output_check==connect_compress
    fprintf('%d:check OK:calc=%d,standard=%d\n',i,connect_compress,output_check);
else
    fprintf('%d:check fail:calc=%d,standard=%d\n',i,connect_compress,output_check);
end

%% close document
fclose(fdata);
fclose(fweight);
fclose(foutput);


% end

File: IC2/sample/check/check.py
# coding:UTF-8
import numpy as np

def str2int8_list(str):
    str_list = str.rstrip().split("\n") # strip last LF; split string to list
    array = [int(x,base=2) for x in str_list] # deal with unsigned number
    array = [x if x<=127 else -((x - 1) ^ int("11111111", base=2)) for x in array] # deal with signed number
    return array

def compress(array):
    array = np.right_shift(array,8)
    array[array > 127] = 127 # use reference!
    array[array < -128] = -128
    return array



# open document
fdata = open("./sample/data.txt","r")
fweight = open("./sample/weight.txt","r")
foutput = open("./sample/output.txt","r")

# preprocess weight
weight_conv = str2int8_list(fweight.read())
weight_connect = np.asarray(weight_conv[3*3*3:],dtype=np.int32).reshape((3,3,3)) # divide into 2 parts
weight_conv =    np.asarray(weight_conv[:3*3*3],dtype=np.int32).reshape((3,3,3))

# preprocess data
data = str2int8_list(fdata.read())
data = [[x for x in data[i*64:(i+1)*64]] for i in range(100)] # divide into 100 pictures
data = [np.asarray(x, dtype=np.int32).reshape((8,8)) for x in data] # convert to ndarray

# preprocess output
output = str2int8_list(foutput.read())
output = [np.asarray(x, dtype=np.int32) for x in output]

for i in range(100):# select pictures

    # calculate convolution
    conv = np.zeros((3,6,6),dtype=np.int32)
    for d in range(3):
        for r in range(6):
            for c in range(6):
                conv[d,r,c] = np.vdot(data[i][r:(r+3),c:(c+3)], weight_conv[d,:,:])

    # get compressed
    conv = compress(conv)

    # calculate pool
    pool = np.zeros((3,3,3), dtype=np.int32)
    # for d in range(3):
    #     for r in range(3):
    #         for c in range(3):
    #             pool[d,r,c] = conv[d,2*r:(2*r+2),2*c:(2*c+2)].max()
    for r in range(3):
        for c in range(3):
            pool[:,[r],[c]] = conv[:,2*r:(2*r+2),2*c:(2*c+2)].max(axis=1).max(axis=1).reshape(3,1)
            # suggest left value to be slice for reference?

    # calculate connect
    connect = np.array([0])
    for d in range(3):
        connect += np.vdot(pool[d,:,:], weight_connect[d,:,:])

    # get compressed
    connect = compress(connect)

    # check answer
    if connect[0] == output[i]:
        print("No.{}:check OK. dout={:d}".format(i, output[i]))
    else:
        print("No.{}:check failed. dout={:d},ans={:d}".format(i, output[i], connect[0]))

# close document
fdata.close()
fweight.close()
foutput.close()
